### DISCUSSION TITLE
Google TV Android 11+ workaround

### DISCUSSION THREAD

**Discussion by lucas-jc** [mupen64plus-ae/mupen64plus-ae#1136] [General]
I finally got my Realme Stick 4k working with the Mupen64Plus by installing the app AnExplorer as my file manager and changing the file picker button method forcing the usage of "startSafFilePicker()" instead of the "startLegacyFilePicker". Now, the emulator uses the AnExplorer to pick the decompressed rom.

Steps:
- Install the file manager AnExplorer (https://play.google.com/store/apps/details?id=dev.dworks.apps.anexplorer) and go to Settings -> Apps -> AnExplorer -> Permissions -> Files and media -> Allow all the time

- Go to the class ScanRomsActivity, go to line 102 where the "filePickerButton" set the click listener, and change the "startFilePicker()" to "startSafFilePicker()"

- Make the project, generate the apk, and install it on the Google TV stick.

- Now the emulator will use the AnExplorer and it has permission to access an external storage.
Note: decompress the rom before opening it. Zip files will not work.

I already generated a debug apk for those who want to test the workaround:
https://drive.google.com/file/d/1GhA-FtQ2mhiOf3ag17OHjgxnGaypSL2q/view?usp=share_link

Tested on: Realme Stick 4k and Android TV API 30 (Android 11.0 Google TV | x86) emulator

**Comment by fzurita** [comment]
Interesting, a more elegant fix would be to split ```useLegacyFileBrowser``` into variables, ```useLegacyFilePicker``` and ```useLegacyFolderPicker``` in here:
https://github.com/mupen64plus-ae/mupen64plus-ae/blob/d0164123a6afb388b3fa46e9525d348b05226ef9/app/src/main/java/paulscode/android/mupen64plusae/persistent/AppData.java#L416-L417

You can set ```useLegacyFolderPicker``` the same way as ```useLegacyFileBrowser``` is initialized. ```useLegacyFilePicker``` would be initialized like this:
```
Intent filePickerIntent= new Intent(Intent.ACTION_OPEN_DOCUMENT);
useLegacyFilePicker = ((isAndroidTv && Build.VERSION.SDK_INT < Build.VERSION_CODES.R) || Build.VERSION.SDK_INT <= Build.VERSION_CODES.O_MR1 ||
                intent.resolveActivity(context.getPackageManager()) == null);
``` 

Then, in here:
https://github.com/mupen64plus-ae/mupen64plus-ae/blob/d0164123a6afb388b3fa46e9525d348b05226ef9/app/src/main/java/paulscode/android/mupen64plusae/ScanRomsActivity.java#L105

Change to:
```
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && appData.useLegacyFilePicker && appData.useLegacyFolderPicker) {
```

And this line:
https://github.com/mupen64plus-ae/mupen64plus-ae/blob/d0164123a6afb388b3fa46e9525d348b05226ef9/app/src/main/java/paulscode/android/mupen64plusae/ScanRomsActivity.java#L134
Change to:
```
if (appData.useLegacyFolderPicker) {
```

And this line:
https://github.com/mupen64plus-ae/mupen64plus-ae/blob/d0164123a6afb388b3fa46e9525d348b05226ef9/app/src/main/java/paulscode/android/mupen64plusae/ScanRomsActivity.java#L143
Change to:
```
if (appData.useLegacyFilePicker) {
```

Of course, there would be a ton of other places where ```useLegacyFileBrowser``` would be referenced. All those would have to be adjusted appropriately, so not the easiest thing to do.

**Comment by fzurita** [comment]
I do wonder if maybe there is an app on the play store that could provide: ```Intent.ACTION_OPEN_DOCUMENT_TREE``` for Android TV as well. The one you found only provides ```Intent.ACTION_OPEN_DOCUMENT```

**Comment by lucas-jc** [comment]
Yes, I totally agree that the workaround was only a first step to a more elegant solution. I post it for those who were desperate like me to get it working 😆  (already bought two bluetooth controllers). Maybe the key is the permission "Allow all the time" in AnExplorer that grants the correct access to the files (special permission in the manifest??).

Thanks in advance for the attention 😄

**Comment by trilkin-github** [comment]
Hi all, 
I finally bought a Chromecast with Google TV, also because I hoped I would be able to play Mario kart on it with this great emulator.
I am also stuck as I cannot open the downloaded ROM. When I click "select file" I get "you don't have an app for that". I downloaded a file browser and thought maybe I can find the location of the app and move the ROM there, but that also didn't work. Is there any workaround that does not include compiling a new APK?

Thanks in advance

**Comment by LeoSum8** [comment]
@lucas-jc do you still happen to have that apk with the modifications you describe?
I'd like to give it a shot but currently can't wrap my head around setting everything up to build it myself

**Comment by lucas-jc** [comment]
Hi @LeoSum8 ,

Of course, you can try this new link https://drive.google.com/file/d/1FkgdEAWDHqdPqM8ttfIzvu1Hc0gIdo3Q/view?usp=drive_link
Remember to install a file manager with the right permissions to access files and use unzipped roms.

Good luck.

**Comment by LeoSum8** [comment]
Thank you this works perfectly!

**Comment by mirko** [comment]
Is there any chance with whatever approach serving this purpose ending up upstream?

**Comment by gdandsm64fan123** [comment]
@lucas-jc, sorry for the interruption but do you happen to still have the APK? i need it for my android tv and would be much appreciated if you could send it again
